### PR TITLE
Add support for spanner table options in import/export

### PR DESCRIPTION
**Pull request by n-d-joshi**
This PR adds: 

1. Support for Import/Export of table options. 
2. Adds an allowlist for the table options which currently only has fulltext_dictionary_table.

### CODE REVIEW

**Comment by n-d-joshi** [comment]
R: @darshan-sj

File: v1/src/main/java/com/google/cloud/teleport/spanner/ddl/InformationSchemaScanner.java [hunk 1]
```diff
@@ -765,6 +766,67 @@ private void listColumnOptions(Ddl.Builder builder) {
...
+      }
+
+      ImmutableList.Builder<String> options =
+          allOptions.computeIfAbsent(tableName, k -> ImmutableList.builder());
+
+      if (optionType.equalsIgnoreCase("STRING")) {
+        String quoteChar =
+            dialect == Dialect.POSTGRESQL ? POSTGRESQL_LITERAL_QUOTE : GSQL_LITERAL_QUOTE;
+        options.add(
+            optionName + "=" + quoteChar + OPTION_STRING_ESCAPER.escape(optionValue) + quoteChar);
```
Comment (darshan-sj) [inline v1/src/main/java/com/google/cloud/teleport/spanner/ddl/InformationSchemaScanner.java]: Why is NameUtils' quoteIdentifier or quoteLiteral not used here? - https://github.com/GoogleCloudPlatform/DataflowTemplates/blob/main/v1/src/main/java/com/google/cloud/teleport/spanner/common/NameUtils.java

File: v1/src/main/java/com/google/cloud/teleport/spanner/ddl/InformationSchemaScanner.java [hunk 2]
```diff
@@ -765,6 +766,67 @@ private void listColumnOptions(Ddl.Builder builder) {
...
+      String tableName = entry.getKey();
+      ImmutableList<String> options = entry.getValue().build();
+      builder.createTable(tableName).tableOptions(options).endTable();
+    }
+  }
+
+  @VisibleForTesting
+  Statement listTableOptionsSQL() {
+    switch (dialect) {
+      case GOOGLE_STANDARD_SQL:
```
Comment (darshan-sj) [inline v1/src/main/java/com/google/cloud/teleport/spanner/ddl/InformationSchemaScanner.java:811]: Are all these columns and tables available in every production databases? Please confirm.

Please get these 2 queries reviewed and LGTMed by one of your team members.

File: v1/src/main/java/com/google/cloud/teleport/spanner/ddl/Table.java [hunk 3]
```diff
@@ -123,6 +126,9 @@ private void prettyPrintPg(
               .collect(Collectors.joining(", ", "\n\tPRIMARY KEY (", ")")));
     }
     appendable.append("\n)");
+    if (!tableOptions().isEmpty()) {
+      appendable.append("\nWITH (").append(String.join(", ", tableOptions())).append(")");
+    }
```
Comment (darshan-sj) [inline v1/src/main/java/com/google/cloud/teleport/spanner/ddl/Table.java]: Is the relative positioning of "with" clause correct in this CREATE TABLE statement?

Also, please add a test case in DdlTest.java covering all the clauses here including Table options.

File: v1/src/main/java/com/google/cloud/teleport/spanner/ddl/Table.java [hunk 4]
```diff
@@ -163,6 +169,9 @@ private void prettyPrintGsql(
               .collect(Collectors.joining(", ", "\n) PRIMARY KEY (", "")));
     }
     appendable.append(")");
+    if (!tableOptions().isEmpty()) {
+      appendable.append(",\nOPTIONS (").append(String.join(", ", tableOptions())).append(")");
+    }
```
Comment (darshan-sj) [inline v1/src/main/java/com/google/cloud/teleport/spanner/ddl/Table.java]: Same comment as PG. I hope the relative ordering of OPTIONS clause is correct here. And Please add test case in DdlTest.java covering all the clauses currently present here.

File: v1/src/main/java/com/google/cloud/teleport/spanner/AvroUtil.java [hunk 5]
```diff
@@ -46,6 +46,7 @@ private AvroUtil() {}
   public static final String SPANNER_PARENT = "spannerParent";
   public static final String SPANNER_INTERLEAVE_TYPE = "spannerInterleaveType";
   public static final String SPANNER_PRIMARY_KEY = "spannerPrimaryKey";
+  public static final String SPANNER_TABLE_OPTIONS = "spannerTableOptions_";
```
Comment (darshan-sj) [inline v1/src/main/java/com/google/cloud/teleport/spanner/AvroUtil.java]: Other numbered properties are singular in this file.

Please change to Singular

SPANNER_TABLE_OPTION = "spannerTableOption_"

File: v1/src/test/java/com/google/cloud/teleport/spanner/AvroSchemaToDdlConverterTest.java [hunk 6]
```diff
@@ -1030,6 +1030,57 @@ public void columnOptions() {
                 + " ) PRIMARY KEY (`id` ASC)"));
   }
 
+  @Test
```
Comment (darshan-sj) [inline v1/src/test/java/com/google/cloud/teleport/spanner/AvroSchemaToDdlConverterTest.java:1033]: I see some escaping is being done for the Table options in InformationSchemaScanner class. Please add test cases where escaping is exercised and escaping is not exercised.

File: v1/src/main/java/com/google/cloud/teleport/spanner/ddl/InformationSchemaScanner.java (hunk 1 above)
Comment (n-d-joshi) [reply v1/src/main/java/com/google/cloud/teleport/spanner/ddl/InformationSchemaScanner.java]: Done. Changed it to NameUtils.

File: v1/src/main/java/com/google/cloud/teleport/spanner/ddl/Table.java (hunk 3 above)
Comment (n-d-joshi) [reply v1/src/main/java/com/google/cloud/teleport/spanner/ddl/Table.java]: I updated to move it after interleave. The rest has some context - will follow up with you offline.

File: v1/src/main/java/com/google/cloud/teleport/spanner/ddl/Table.java (hunk 4 above)
Comment (n-d-joshi) [reply v1/src/main/java/com/google/cloud/teleport/spanner/ddl/Table.java]: Moved it after interleave in and added a test.

File: v1/src/main/java/com/google/cloud/teleport/spanner/AvroUtil.java (hunk 5 above)
Comment (n-d-joshi) [reply v1/src/main/java/com/google/cloud/teleport/spanner/AvroUtil.java]: Done.

File: v1/src/test/java/com/google/cloud/teleport/spanner/AvroSchemaToDdlConverterTest.java (hunk 6 above)
Comment (n-d-joshi) [reply v1/src/test/java/com/google/cloud/teleport/spanner/AvroSchemaToDdlConverterTest.java:1033]: I am following the same convention as the other tests in creating the table:

            "CREATE TABLE `CustomDictionary` ("
                + " `Key`                                    STRING(MAX) NOT NULL,"
                + " `Value`                                  ARRAY<STRING(MAX)> NOT NULL,"
                + " ) PRIMARY KEY (`Key` ASC),\nOPTIONS (fulltext_dictionary_table=true)"));

This is not escaped? 
Did I misunderstand the comment?

File: v1/src/main/java/com/google/cloud/teleport/spanner/ddl/InformationSchemaScanner.java [hunk 7]
```diff
@@ -765,6 +766,66 @@ private void listColumnOptions(Ddl.Builder builder) {
...
+
+      ImmutableList.Builder<String> options =
+          allOptions.computeIfAbsent(tableName, k -> ImmutableList.builder());
+
+      if (optionType.equalsIgnoreCase("STRING")) {
+        String quoteChar = com.google.cloud.teleport.spanner.common.NameUtils.literalQuote(dialect);
+        options.add(
+            optionName + "=" + quoteChar + OPTION_STRING_ESCAPER.escape(optionValue) + quoteChar);
+      } else if (optionType.equalsIgnoreCase("character varying")) {
+        options.add(optionName + "='" + OPTION_STRING_ESCAPER.escape(optionValue) + "'");
```
Comment (darshan-sj) [inline v1/src/main/java/com/google/cloud/teleport/spanner/ddl/InformationSchemaScanner.java:795]: I'm referring to this option_string_escaper string in the comment in AvroSchemaToDdlConverterTest.java‎ . Does the test case added there test different edge cases around this escaping?

File: v1/src/main/java/com/google/cloud/teleport/spanner/ddl/Table.java (hunk 3 above)
Comment (n-d-joshi) [reply v1/src/main/java/com/google/cloud/teleport/spanner/ddl/Table.java]: As discussed offline - this PR is only to add the table options support for dictionary tables. dictionary tables cannot have any other clause other than OPTIONS.

So all clauses with Table options can be done in a follow up CL when other table options are added.

File: v1/src/main/java/com/google/cloud/teleport/spanner/ddl/InformationSchemaScanner.java (hunk 2 above)
Comment (xinbei-liu) [reply v1/src/main/java/com/google/cloud/teleport/spanner/ddl/InformationSchemaScanner.java:811]: Yes this should be available in all database as long as we expose the table_options table.
